sql/src/main/java/io/crate/planner/operators/LogicalPlanner.java

```diff
@@ -0,0 +1,178 @@
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.operators;
+
+import io.crate.analyze.OrderBy;
+import io.crate.analyze.QueriedTableRelation;
+import io.crate.analyze.QueryClause;
+import io.crate.analyze.WhereClause;
+import io.crate.analyze.relations.QueriedRelation;
+import io.crate.analyze.symbol.FieldsVisitor;
+import io.crate.analyze.symbol.Function;
+import io.crate.analyze.symbol.Literal;
+import io.crate.analyze.symbol.RefVisitor;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.planner.MultiPhasePlan;
+import io.crate.planner.Plan;
+import io.crate.planner.Planner;
+import io.crate.planner.SubqueryPlanner;
+import io.crate.planner.projection.builder.ProjectionBuilder;
+import io.crate.planner.projection.builder.SplitPoints;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+/**
+ * Planner which can create a {@link Plan} using intermediate {@link LogicalPlan} nodes.
+ */
+public class LogicalPlanner {
+
+    public static final int NO_LIMIT = -1;
+
+    public Plan plan(QueriedRelation queriedRelation,
+                     Planner.Context plannerContext,
+                     ProjectionBuilder projectionBuilder) {
+        LogicalPlan logicalPlan = plan(queriedRelation)
+            .build(extractColumns(queriedRelation.querySpec().outputs()))
+            .tryCollapse();
+
+        Plan plan = logicalPlan.build(
+            plannerContext,
+            projectionBuilder,
+            LogicalPlanner.NO_LIMIT,
+            0,
+            null
+        );
+        SubqueryPlanner subqueryPlanner = new SubqueryPlanner(plannerContext);
+        return MultiPhasePlan.createIfNeeded(plan, subqueryPlanner.planSubQueries(queriedRelation.querySpec()));
+    }
+
+    static LogicalPlan.Builder plan(QueriedRelation relation) {
+        SplitPoints splitPoints = SplitPoints.create(relation.querySpec());
+        return fetchOrEval(
+            limit(
+                orderBy(
+                    filter(
+                        groupByOrAggregate(
+                            whereAndCollect(relation, splitPoints.toCollect(), relation.where()),
```

Comment:
That's kind of nice. Maybe you could also indent `whereAndCollect` like you did with the others.